<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Extract instruments locally for MR Analysis — extract_instruments_local • leo.gwas</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extract instruments locally for MR Analysis — extract_instruments_local"><meta name="description" content="Filters SNPs by p-value threshold and performs LD clumping with flexible column mapping."><meta property="og:description" content="Filters SNPs by p-value threshold and performs LD clumping with flexible column mapping."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">leo.gwas</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/installation.html">Installation Tutorial</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/laleoarrow/leo.gwas" aria-label="GitHub repository"><span class="fa fa-github"></span></a></li>
<li class="nav-item"><a class="nav-link" href="https://laleoarrow.github.io/leo.gwas/" aria-label="Package home"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Extract instruments locally for MR Analysis</h1>

      <div class="d-none name"><code>extract_instruments_local.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Filters SNPs by p-value threshold and performs LD clumping with flexible column mapping.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extract_instruments_local</span><span class="op">(</span></span>
<span>  <span class="va">dat</span>,</span>
<span>  p <span class="op">=</span> <span class="fl">5e-08</span>,</span>
<span>  pop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phenotype_col <span class="op">=</span> <span class="st">"Phenotype"</span>,</span>
<span>  snp_col <span class="op">=</span> <span class="st">"SNP"</span>,</span>
<span>  chr_col <span class="op">=</span> <span class="st">"CHR"</span>,</span>
<span>  pos_col <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>  effect_allele_col <span class="op">=</span> <span class="st">"A1"</span>,</span>
<span>  other_allele_col <span class="op">=</span> <span class="st">"A2"</span>,</span>
<span>  beta_col <span class="op">=</span> <span class="st">"BETA"</span>,</span>
<span>  se_col <span class="op">=</span> <span class="st">"SE"</span>,</span>
<span>  pval_col <span class="op">=</span> <span class="st">"P"</span>,</span>
<span>  eaf_col <span class="op">=</span> <span class="st">"EAF"</span>,</span>
<span>  N <span class="op">=</span> <span class="st">"Neff"</span>,</span>
<span>  bfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plink_bin <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dat">dat<a class="anchor" aria-label="anchor" href="#arg-dat"></a></dt>
<dd><p>Dataframe containing GWAS summary statistics; change it to dataframe if it is data.table.</p></dd>


<dt id="arg-p">p<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>P-value cutoff (default = 5e-8)</p></dd>


<dt id="arg-pop">pop<a class="anchor" aria-label="anchor" href="#arg-pop"></a></dt>
<dd><p>Super-population code for LD reference (default = NULL, as we recommend using loca bfile)</p></dd>


<dt id="arg-phenotype-col">phenotype_col<a class="anchor" aria-label="anchor" href="#arg-phenotype-col"></a></dt>
<dd><p>Column name for phenotype (default = "Phenotype")</p></dd>


<dt id="arg-snp-col">snp_col<a class="anchor" aria-label="anchor" href="#arg-snp-col"></a></dt>
<dd><p>Column name for SNP IDs (default = "SNP")</p></dd>


<dt id="arg-chr-col">chr_col<a class="anchor" aria-label="anchor" href="#arg-chr-col"></a></dt>
<dd><p>Column name for chromosome (default = "CHR")</p></dd>


<dt id="arg-pos-col">pos_col<a class="anchor" aria-label="anchor" href="#arg-pos-col"></a></dt>
<dd><p>Column name for position (default = "POS")</p></dd>


<dt id="arg-effect-allele-col">effect_allele_col<a class="anchor" aria-label="anchor" href="#arg-effect-allele-col"></a></dt>
<dd><p>Column name for effect allele (default = "A1")</p></dd>


<dt id="arg-other-allele-col">other_allele_col<a class="anchor" aria-label="anchor" href="#arg-other-allele-col"></a></dt>
<dd><p>Column name for non-effect allele (default = "A2")</p></dd>


<dt id="arg-beta-col">beta_col<a class="anchor" aria-label="anchor" href="#arg-beta-col"></a></dt>
<dd><p>Column name for effect size (default = "BETA")</p></dd>


<dt id="arg-se-col">se_col<a class="anchor" aria-label="anchor" href="#arg-se-col"></a></dt>
<dd><p>Column name for standard error (default = "SE")</p></dd>


<dt id="arg-pval-col">pval_col<a class="anchor" aria-label="anchor" href="#arg-pval-col"></a></dt>
<dd><p>Column name for p-values (default = "P")</p></dd>


<dt id="arg-eaf-col">eaf_col<a class="anchor" aria-label="anchor" href="#arg-eaf-col"></a></dt>
<dd><p>Column name for effect allele frequency (default = "EAF")</p></dd>


<dt id="arg-n">N<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>Column name for sample size (default = "Neff", set NULL to exclude)</p></dd>


<dt id="arg-bfile">bfile<a class="anchor" aria-label="anchor" href="#arg-bfile"></a></dt>
<dd><p>Path to PLINK binary reference panel (default = NULL)</p></dd>


<dt id="arg-plink-bin">plink_bin<a class="anchor" aria-label="anchor" href="#arg-plink-bin"></a></dt>
<dd><p>Path to PLINK executable (default = NULL)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Formatted exposure data ready for MR analysis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Custom column names example</span></span></span>
<span class="r-in"><span><span class="fu">extract_instruments_local</span><span class="op">(</span></span></span>
<span class="r-in"><span>  dat <span class="op">=</span> <span class="va">gwas_data</span>,</span></span>
<span class="r-in"><span>  phenotype_col <span class="op">=</span> <span class="st">"Trait"</span>,</span></span>
<span class="r-in"><span>  snp_col <span class="op">=</span> <span class="st">"rsID"</span>,</span></span>
<span class="r-in"><span>  chr_col <span class="op">=</span> <span class="st">"Chromosome"</span>,</span></span>
<span class="r-in"><span>  pos_col <span class="op">=</span> <span class="st">"Position"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ao Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

